binlog_reconciler.go

```diff
@@ -0,0 +1,223 @@
+			" VALUES (" + buildStringListForValues(e.newValues) + ")"
+	} else {
+		pkColumnName := e.TableSchema().GetPKColumn(0).Name
+		if pkColumnName == "" {
+			return "", fmt.Errorf("cannot get PK column for table %s", e.Table())
+		}
+
+		pkColumnName = quoteField(pkColumnName)
+		query = "DELETE FROM " + QuotedTableNameFromString(target.Schema, target.Name) +
+			" WHERE " + buildStringMapForWhere([]string{pkColumnName}, []interface{}{e.pk})
+	}
+
+	return query, nil
+}
+
+func NewReconciliationDMLEvent(table *schema.Table, pk uint64, row RowData) DMLEvent {
+	return &ReconcilationDMLEvent{
+		DMLEventBase: &DMLEventBase{table: *table},
+		pk:           pk,
+		newValues:    row,
+	}
+}
+
+// Instead of replacing/deleting every row, we first store all the rows changed
+// during the downtime and perform all the operations at the end. This maybe
+// faster (trading memory usage tho) and it is crucial for schema changes (as
+// we can simply delete a key from this map when we realize a table has
+// changed).
+type UniqueRowMap map[TableIdentifier]map[uint64]struct{}
+
+func (m UniqueRowMap) AddRow(table *schema.Table, pk uint64) bool {
+	tableId := NewTableIdentifierFromSchemaTable(table)
+	if _, exists := m[tableId]; !exists {
+		m[tableId] = make(map[uint64]struct{})
+	}
+
+	if _, exists := m[tableId][pk]; !exists {
+		m[tableId][pk] = struct{}{}
+		return true
+	}
+
+	return false
+}
+
+type BinlogReconciler struct {
+	BatchSize        int
+	TableSchemaCache TableSchemaCache
+
+	SourceDB     *sql.DB
+	BinlogWriter *BinlogWriter
+
+	modifiedRows UniqueRowMap
+	logger       *logrus.Entry
+}
+
+func (r *BinlogReconciler) Initialize() {
+	r.modifiedRows = make(UniqueRowMap)
+	r.logger = logrus.WithField("tag", "binlogreconcile")
+}
+
+func (r *BinlogReconciler) AddRowsToStore(events []DMLEvent) error {
+	for _, ev := range events {
+		pk, err := ev.PK()
+		if err != nil {
+			return err
+		}
+
+		r.modifiedRows.AddRow(ev.TableSchema(), pk)
+	}
+
+	return nil
+}
+
+func (r *BinlogReconciler) ReplaceModifiedRowsAfterCatchup() error {
+	batch := make([]DMLEvent, 0, r.BatchSize)
+
+	totalModifiedRows := 0
+	for _, pkSet := range r.modifiedRows {
+		totalModifiedRows += len(pkSet)
+	}
+
+	r.logger.WithField("row_count", totalModifiedRows).Info("begin replacing modified rows")
+
+	count := 0
+	for tableId, pkSet := range r.modifiedRows {
+		table := r.TableSchemaCache.Get(tableId.SchemaName, tableId.TableName)
+
+		for pk, _ := range pkSet {
+			count++
+
+			if len(batch) == r.BatchSize {
+				r.logger.WithField("rows_replaced", count).Debug("replacing batch")
+				err := r.replaceBatch(batch)
+				if err != nil {
+					return err
+				}
+
+				batch = make([]DMLEvent, 0, r.BatchSize)
+			}
+
+			row, err := r.selectRowFromSource(table, pk)
+			if err != nil {
+				r.logger.WithError(err).Error("failed to select row from source")
+				return err
+			}
+
+			batch = append(batch, NewReconciliationDMLEvent(table, pk, row))
```

Comment:
wait... why do you need `ReconcilationDMLEvent` at all? can't you re-use `RowBatch` here?